subsys/pm/device_runtime.c

```diff
@@ -28,13 +28,18 @@ LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
  * @retval -ENOSTUP If runtime PM is not enabled for the device.
  * @retval -EALREADY If device is already suspended (can only happen if get/put
  * calls are unbalanced).
+ * @retval -EPERM If the device has a pm constraint.
  * @retval -errno Other negative errno, result of the action callback.
  */
 static int runtime_suspend(const struct device *dev, bool async)
 {
 	int ret = 0;
 	struct pm_device *pm = dev->pm;
 
+	if (pm_device_constraint_check(dev)) {
+		return -EPERM;
+	}
```

Comment:
I'm a bit skeptical about this change, since it makes drivers error handling more complicated. If a driver now does:

```
ret = pm_device_runtime_put(dev);
if (ret < 0) {
    return ret;
}
```

it will return an error. However, the driver operation will likely be ok (just that PM suspend/resume is not allowed at this time). To summarize, unless driver ignores `-EPERM`, it will become unusable. A similar situation happens in other scenarios, for which I've created https://github.com/zephyrproject-rtos/zephyr/pull/40670.

Reply:
Another point: for runtime operation the device usage count may still be updated (but no action taken). I'm worried about introducing a mechanism that can potentially interfere with the runtime API usage, potentially leaving the system into an inconsistent state.

Reply:
I thought about it, my opinion *now* is that to allow  increase/decrease the reference count when it does not trigger a state change makes this API behavior trickier.
My initial idea was to allow use it only when device runtime is disabled, the problem is that when runtime pm is disabled the device is always resumed, consequently if one wants to set a constraint (or lock) the device when the state is SUSPENDED it will be necessary to use `pm_device_action_run` and we don't want this API directly called. Though it makes the behavior much more predictable and the only `pm_device_runtime_enable` should be changed to return `-EPERM` when the device has a constraint (or is locked)